File: ProductKey/WSH/source.js
var ProductKey = ProductKey || {
  DecodeProductKey : function(digitalProductId) {
    var map = ('BCDFGHJKMPQRTVWXY2346789').split(''),
        key = [], raw, i, j;
        
    with (new ActiveXObject('WScript.Shell'))
      raw = RegRead(digitalProductId).toArray().slice(52, 67);
      
    for (i = 24; i >= 0; i--) {
      var k = 0;
      for (j = 14; j >= 0; j--) {
        k = (k * 256) ^ raw[j];
        raw[j] = Math.floor(k / 24);
        k %= 24;
      }
      key = map[k] + key;
      
      if ((i % 5) == 0 && i != 0) key = '-' + key;
    }
    return key;
  },
  
  GetProductKey : function() {
    var key = 'HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\DigitalProductId';
    return this.DecodeProductKey(key);
  }
};

WScript.echo(ProductKey.GetProductKey());

File: ProductKey/NET/source.rb
require 'System'

Registry = Microsoft::Win32::Registry

def DecodeProductKey(digitalProductId)
  _map = ("BCDFGHJKMPQRTVWXY2346789").split('')
  _key = []
  _raw = []
  
  for i in 52..67 do
    _raw.Add(digitalProductId[i])
  end
  
  i = 28
  while i >= 0
    if (i + 1) % 6 == 0
      _key[i] = '-'
    else
      k = 0
      j = 14
      while j >= 0
        k = (k * 256) ^ _raw[j]
        _raw[j] = k / 24
        k %= 24
        _key[i] = _map[k]
        j -= 1
      end
    end
    i -= 1
  end
  
  return _key.to_s
end

rk = Registry.LocalMachine.OpenSubKey('SOFTWARE\Microsoft\Windows NT\CurrentVersion')
puts DecodeProductKey(rk.GetValue('DigitalProductId'))
rk.Close()
